src/node_crypto.cc

```diff
@@ -1025,6 +1066,46 @@ void SecureContext::LoadPKCS12(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+#ifndef OPENSSL_NO_ENGINE
+void SecureContext::SetClientCertEngine(
+  const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK(args.Length() == 1 && args[0]->IsString());
+
+  SecureContext* sc = Unwrap<SecureContext>(args.This());
+
+  ClearErrorOnReturn clear_error_on_return;
+
+  // If an engine was previously set, free it.
+  // This is because SSL_CTX_set_client_cert_engine does not itself
+  // 'support' multiple calls by cleaning up before overwriting the
+  // client_cert_engine internal context variable
+  if (sc->ctx_->client_cert_engine != NULL) {
```

Comment:
It's a good question. It's a bug I believe in OpenSSL that SSL_CTX_set_client_cert_engine does not clean up a previous set engine by calling ENGINE_finish. We could maintain an explicit property that can tell us if we have previously set an engine - but if OpenSSL suddenly fixes their issue it would still be wrong to ENGINE_finish the engine (as OpenSSL would do this)... I came up with a solution that I think has the greatest chance of not breaking in the future. If it breaks it would be because OpenSSL changed their internal variable name but in that case it would cause a build error which is to be preferred over a run-time crash due to double ENGINE_finish.

Also given my git problems I would really love not having to continue work on this pull-request ;-)